x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/DelimitedFileStructureFinder.java

```diff
@@ -529,11 +535,22 @@ static boolean canCreateFromSample(List<String> explanation, String sample, int
                         --fieldsInThisRow;
                     }
 
-                    if (fieldsInLastRow != fieldsInFirstRow) {
-                        explanation.add("Not " + formatName + " because row [" + (numberOfRows - 1) +
-                            "] has a different number of fields to the first row: [" + fieldsInFirstRow + "] and [" +
-                            fieldsInLastRow + "]");
-                        return false;
+                    // TODO: might be good one day to gather a distribution of the most common field counts
+                    // But, this would require iterating (or at least sampling) all the lines.
+                    if (fieldsInThisRow != fieldsInFirstRow) {
+                        illFormattedRows.add(numberOfRows - 1);
+                        // We should only allow a certain percentage of ill formatted rows
+                        // as it may effects our sample sizing and down stream processing
+                        if (illFormattedRows.size() > Math.ceil(allowedFractionOfBadLines * numberOfRows)) {
```

Comment:
I think there's a problem here that I didn't think of when I suggested to move the check inside the loop.

If the bad row comes very early on then this will trigger when the overall proportion of bad lines in the file is within the tolerance.

You should be able to reproduce this by making your bad row the second non-header row in your test case.  At the point it occurs 50% of rows will be bad.

So, we need a formula that will short-circuit as soon as it's impossible for the overall percentage to be within the tolerance (which will be on the first bad row in the no overrides case) but not sooner.  Unfortunately we don't know the number of CSV rows for sure because some could be multi-line.  Assuming worst case, let's make the formula work on the basis that every subsequent sample line is one row, and they're all good.  That leads to:

```suggestion
                        if (illFormattedRows.size() > Math.ceil(allowedFractionOfBadLines * (numberOfRows + numberOfLinesInSample - csvReader.getLineNumber()))) {
```

There might be an off-by-one error in that formula depending on exactly what `csvReader.getLineNumber()` after reading a row.

The next problem is that `numberOfLinesInSample` isn't known in this method.  `FileStructureFinderManager.sampleFile()` knows it, so you could change that to return it and pass it through to the `canCreate___` methods, or maybe there is somewhere else it's more convenient to get it from.

Reply:
AGH! You are right. Need to adjust

Reply:
@droberts195 looking at the delimited factory. I don't see `lineMergeSizeLimit` used at all. And following what is done in `FileStructureFinderManager.sampleFile` it SEEMS to simply read lines and split...which is the same thing that is done in https://github.com/elastic/elasticsearch/blob/257b9a69cca370666a4712e4267be837c099428a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/DelimitedFileStructureFinder.java#L495

I might could use `sampleLines.length` as an approximation.

Reply:
> looking at the delimited factory. I don't see `lineMergeSizeLimit` used at all

Good point, that's a bug.  It can be fixed by changing `DelimitedFileStructureFinderFactory.createFromSample` to this:

```
        CsvPreference adjustedCsvPreference = new CsvPreference.Builder(csvPreference).maxLinesPerRow(lineMergeSizeLimit).build();
        return DelimitedFileStructureFinder.makeDelimitedFileStructureFinder(explanation, sample, charsetName, hasByteOrderMarker,
            adjustedCsvPreference, trimFields, overrides, timeoutChecker);
```

That can be done in this PR or a subsequent one, as the problem is independent of the main change of this PR.

> I might could use `sampleLines.length` as an approximation.

Yes, true, I missed that.